## ENDEAVOUR INDONESIA adalah sebuah agensi yang bertugas untuk melakukan assesmen terhadap suatu kasus berdasarkan data yang sudah tersedia.

#### Pada kesempatan kali ini, agensi mendapatkan proyek untuk menilai apakah seorang pemain sepakbola dunia layak untuk mendapatkan nilai gaji tertentu berdasarkan parameter-parameter terkait.

#### Agensi akan mengolah data yang diperoleh dari kaggle.com dengan judul dataset FIFA 22 Players Dataset

In [786]:
import pandas as pd

In [787]:
data = pd.read_csv(r"C:\Users\CPU\Downloads\FIFA Dataset\players_22.csv",low_memory=False)

In [788]:
data.head()

sofifa_id                                         player_url  \
0     158023  https://sofifa.com/player/158023/lionel-messi/...   
1     188545  https://sofifa.com/player/188545/robert-lewand...   
2      20801  https://sofifa.com/player/20801/c-ronaldo-dos-...   
3     190871  https://sofifa.com/player/190871/neymar-da-sil...   
4     192985  https://sofifa.com/player/192985/kevin-de-bruy...   

          short_name                            long_name player_positions  \
0           L. Messi       Lionel Andrés Messi Cuccittini       RW, ST, CF   
1     R. Lewandowski                   Robert Lewandowski               ST   
2  Cristiano Ronaldo  Cristiano Ronaldo dos Santos Aveiro           ST, LW   
3          Neymar Jr        Neymar da Silva Santos Júnior          LW, CAM   
4       K. De Bruyne                      Kevin De Bruyne          CM, CAM   

   overall  potential    value_eur  wage_eur  age  ...   lcb    cb   rcb  \
0       93         93   78000000.0  320000.0   34  ...  50+3  50+3  50+3   
1       92         92  119500000.0  270000.0   32  ...  60+3  60+3  60+3   
2       91         91   45000000.0  270000.0   36  ...  53+3  53+3  53+3   
3       91         91  129000000.0  270000.0   29  ...  50+3  50+3  50+3   
4       91         91  125500000.0  350000.0   30  ...  69+3  69+3  69+3   

     rb    gk                                    player_face_url  \
0  61+3  19+3  https://cdn.sofifa.net/players/158/023/22_120.png   
1  61+3  19+3  https://cdn.sofifa.net/players/188/545/22_120.png   
2  60+3  20+3  https://cdn.sofifa.net/players/020/801/22_120.png   
3  62+3  20+3  https://cdn.sofifa.net/players/190/871/22_120.png   
4  75+3  21+3  https://cdn.sofifa.net/players/192/985/22_120.png   

                            club_logo_url  \
0  https://cdn.sofifa.net/teams/73/60.png   
1  https://cdn.sofifa.net/teams/21/60.png   
2  https://cdn.sofifa.net/teams/11/60.png   
3  https://cdn.sofifa.net/teams/73/60.png   
4  https://cdn.sofifa.net/teams/10/60.png   

                             club_flag_url  \
0      https://cdn.sofifa.net/flags/fr.png   
1      https://cdn.sofifa.net/flags/de.png   
2  https://cdn.sofifa.net/flags/gb-eng.png   
3      https://cdn.sofifa.net/flags/fr.png   
4  https://cdn.sofifa.net/flags/gb-eng.png   

                            nation_logo_url  \
0  https://cdn.sofifa.net/teams/1369/60.png   
1  https://cdn.sofifa.net/teams/1353/60.png   
2  https://cdn.sofifa.net/teams/1354/60.png   
3                                       NaN   
4  https://cdn.sofifa.net/teams/1325/60.png   

                       nation_flag_url  
0  https://cdn.sofifa.net/flags/ar.png  
1  https://cdn.sofifa.net/flags/pl.png  
2  https://cdn.sofifa.net/flags/pt.png  
3  https://cdn.sofifa.net/flags/br.png  
4  https://cdn.sofifa.net/flags/be.png  

[5 rows x 110 columns]

##### Selanjutnya, dicari nilai korelasi antar variabel

In [789]:
correl = data.corr()

In [790]:
correl

sofifa_id   overall  potential  value_eur  \
sofifa_id                     1.000000 -0.491523   0.037576  -0.128102   
overall                      -0.491523  1.000000   0.644275   0.554684   
potential                     0.037576  0.644275   1.000000   0.527738   
value_eur                    -0.128102  0.554684   0.527738   1.000000   
wage_eur                     -0.253044  0.601764   0.497611   0.823527   
age                          -0.781023  0.459451  -0.264142   0.042886   
height_cm                    -0.100362  0.042787   0.004403   0.010291   
weight_kg                    -0.207461  0.150324  -0.016912   0.035188   
club_team_id                  0.110305 -0.226640  -0.330928  -0.191468   
league_level                 -0.036769 -0.179032  -0.153075  -0.117917   
club_jersey_number            0.183841 -0.186166   0.009784  -0.067670   
club_contract_valid_until     0.128058  0.184987   0.345188   0.217051   
nationality_id                0.065832 -0.168315  -0.247198  -0.083983   
nation_team_id               -0.009138 -0.214303  -0.185054  -0.084325   
nation_jersey_number          0.179041 -0.116020  -0.027984  -0.017193   
weak_foot                    -0.121359  0.223762   0.157053   0.149757   
skill_moves                  -0.139709  0.376226   0.283746   0.257232   
international_reputation     -0.334292  0.470647   0.357283   0.632181   
release_clause_eur           -0.113962  0.545197   0.537715   0.995460   
pace                          0.162743  0.174695   0.258817   0.181934   
shooting                     -0.301698  0.489623   0.287862   0.290294   
passing                      -0.412920  0.715001   0.437655   0.404774   
dribbling                    -0.296408  0.666402   0.490270   0.404812   
defending                    -0.234020  0.346760   0.162151   0.147359   
physic                       -0.403516  0.529234   0.155740   0.215105   
attacking_crossing           -0.166999  0.401356   0.243757   0.227414   
attacking_finishing          -0.126239  0.338746   0.230814   0.217691   
attacking_heading_accuracy   -0.149084  0.343901   0.192825   0.165921   
attacking_short_passing      -0.183969  0.516788   0.367162   0.297419   
attacking_volleys            -0.197351  0.380942   0.235766   0.238366   
skill_dribbling              -0.085064  0.384351   0.311396   0.240916   
skill_curve                  -0.212534  0.420423   0.270217   0.252820   
skill_fk_accuracy            -0.230161  0.381030   0.207769   0.214721   
skill_long_passing           -0.229260  0.498216   0.324791   0.278721   
skill_ball_control           -0.142978  0.460411   0.347172   0.269854   
movement_acceleration         0.103243  0.194044   0.232621   0.157690   
movement_sprint_speed         0.101162  0.205736   0.235577   0.164682   
movement_agility             -0.059928  0.270279   0.223385   0.176597   
movement_reactions           -0.482973  0.871823   0.522483   0.492741   
movement_balance              0.015254  0.123700   0.136890   0.117887   
power_shot_power             -0.327861  0.561180   0.332292   0.301723   
power_jumping                -0.222044  0.284689   0.096992   0.118294   
power_stamina                -0.115509  0.385150   0.199743   0.198771   
power_strength               -0.305497  0.364366   0.073567   0.128684   
power_long_shots             -0.203978  0.410010   0.243294   0.231585   
mentality_aggression         -0.246133  0.400161   0.191501   0.183251   
mentality_interceptions      -0.173451  0.316470   0.157728   0.141060   
mentality_positioning        -0.135519  0.358562   0.233070   0.224097   
mentality_vision             -0.267625  0.523025   0.345298   0.324453   
mentality_penalties          -0.182124  0.343690   0.208062   0.196231   
mentality_composure          -0.420836  0.708867   0.426790   0.389065   
defending_marking_awareness  -0.161311  0.312093   0.161288   0.140981   
defending_standing_tackle    -0.094413  0.259773   0.156649   0.120724   
defending_sliding_tackle     -0.0

##### Karena yang menjadi fokus adalah masalah gaji, jadi yang perlu dilihat adalah nilai korelasi variabel 'wage_eur' dengan variabel-variabel yang memiliki nilai korelasi mendekati 1 yang berarti berkorelasi positif (dalam hal ini nilai korelasi yang diinginkan adalah lebih dari 0.6)

In [791]:
data1 = pd.DataFrame(correl['wage_eur']).copy()

In [792]:
data1 = data1.loc[(correl['wage_eur'])>0.6].copy()

In [793]:
data1

wage_eur
overall                   0.601764
value_eur                 0.823527
wage_eur                  1.000000
international_reputation  0.675637
release_clause_eur        0.821330

##### Setelah didapatkan variabel yang berkolerasidengan variabel wage_eur, maka selanjutnya akan dicari nilai korelasi variabel tersebut dengan variabel lainnya yang mendukung besaran nilai pada variabel tersebut.

In [794]:
data2 = pd.DataFrame(correl['overall']).copy()

In [795]:
data2 = data2.loc[(correl['overall'])>0.6].copy()

In [796]:
data2

overall
overall              1.000000
potential            0.644275
wage_eur             0.601764
passing              0.715001
dribbling            0.666402
movement_reactions   0.871823
mentality_composure  0.708867

In [797]:
data3 = pd.DataFrame(correl['value_eur']).copy()

In [798]:
data3 = data3.loc[(correl['value_eur'])>0.6].copy()

In [799]:
data3

value_eur
value_eur                  1.000000
wage_eur                   0.823527
international_reputation   0.632181
release_clause_eur         0.995460

In [800]:
data4 = pd.DataFrame(correl['international_reputation']).copy()

In [801]:
data4 = data4.loc[(correl['international_reputation'])>0.6].copy()

In [802]:
data4

international_reputation
value_eur                                 0.632181
wage_eur                                  0.675637
international_reputation                  1.000000
release_clause_eur                        0.615685

In [803]:
data5 = pd.DataFrame(correl['release_clause_eur']).copy()

In [804]:
data5 = data5.loc[(correl['release_clause_eur'])>0.6].copy()

In [805]:
data5

release_clause_eur
value_eur                           0.995460
wage_eur                            0.821330
international_reputation            0.615685
release_clause_eur                  1.000000

#### Variabel 'value_eur', 'international_reputation', dan 'release_clause_eur' hanya berkorelasi dengan variabel diantara mereka sendiri dan dengan variabel 'wage_eur' dan 'overall'. Sedangkan variabel 'overall' memiliki korelasi positif yang cukup tinggi dengan variabel-variabel lain, sehingga akan dicari nilai asumsi dari variabel 'overall' dengan menggunakan variabel-variabel tersebut, kecuali variabel-variabel yang sudah disebutkan di atas tadi.

In [806]:
dataset1 = data[['short_name','overall','passing','dribbling','movement_reactions','mentality_composure']]
dataset1

short_name  overall  passing  dribbling  movement_reactions  \
0                L. Messi       93     91.0       95.0                  94   
1          R. Lewandowski       92     79.0       86.0                  93   
2       Cristiano Ronaldo       91     80.0       88.0                  94   
3               Neymar Jr       91     86.0       94.0                  89   
4            K. De Bruyne       91     93.0       88.0                  91   
...                   ...      ...      ...        ...                 ...   
19234           Song Defu       47     46.0       48.0                  53   
19235           C. Porter       47     50.0       46.0                  49   
19236            N. Logue       47     45.0       49.0                  46   
19237           L. Rudden       47     36.0       48.0                  48   
19238  E. Lalchhanchhuaha       47     45.0       48.0                  54   

       mentality_composure  
0                       96  
1                       88  
2                       95  
3                       93  
4                       89  
...                    ...  
19234                   37  
19235                   47  
19236                   36  
19237                   47  
19238                   36  

[19239 rows x 6 columns]

In [807]:
training_dataset1 = dataset1.loc[(dataset1['overall'])<80].copy()
training_dataset1 = training_dataset1.dropna()
training_dataset1

short_name  overall  passing  dribbling  movement_reactions  \
509               Iniesta       79     84.0       84.0                  71   
510             J. Milner       79     81.0       77.0                  84   
511             F. Ribéry       79     81.0       85.0                  69   
513              B. Gomis       79     68.0       72.0                  80   
514            M. Dembélé       79     76.0       85.0                  71   
...                   ...      ...      ...        ...                 ...   
19234           Song Defu       47     46.0       48.0                  53   
19235           C. Porter       47     50.0       46.0                  49   
19236            N. Logue       47     45.0       49.0                  46   
19237           L. Rudden       47     36.0       48.0                  48   
19238  E. Lalchhanchhuaha       47     45.0       48.0                  54   

       mentality_composure  
509                     89  
510                     84  
511                     81  
513                     81  
514                     90  
...                    ...  
19234                   37  
19235                   47  
19236                   36  
19237                   47  
19238                   36  

[16668 rows x 6 columns]

In [808]:
test_dataset1 = dataset1.loc[(dataset1['overall'])>=80].copy()
test_dataset1 = test_dataset1.dropna()
test_dataset1

short_name  overall  passing  dribbling  movement_reactions  \
0             L. Messi       93     91.0       95.0                  94   
1       R. Lewandowski       92     79.0       86.0                  93   
2    Cristiano Ronaldo       91     80.0       88.0                  94   
3            Neymar Jr       91     86.0       94.0                  89   
4         K. De Bruyne       91     93.0       88.0                  91   
..                 ...      ...      ...        ...                 ...   
504          N. Vlašić       80     78.0       82.0                  78   
505        Pedro Porro       80     73.0       79.0                  78   
506       V. Tsygankov       80     78.0       80.0                  78   
507          V. Coufal       80     73.0       75.0                  78   
508            B. Saka       80     76.0       82.0                  76   

     mentality_composure  
0                     96  
1                     88  
2                     95  
3                     93  
4                     89  
..                   ...  
504                   77  
505                   78  
506                   76  
507                   72  
508                   73  

[439 rows x 6 columns]

In [809]:
X_train = training_dataset1[['passing','dribbling','movement_reactions','mentality_composure']]
y_train = training_dataset1[['overall']]

In [810]:
X_test = test_dataset1[['passing','dribbling','movement_reactions','mentality_composure']]
y_test = test_dataset1[['overall']]

##### Menggunakan algoritma Decision Tree Clasiifier

In [811]:
from sklearn.tree import DecisionTreeRegressor
model = DecisionTreeRegressor()
model = model.fit(X_train, y_train)

In [812]:
y_pred = model.predict(X_test)

In [813]:
y_pred1 = pd.Series(y_pred)
y_pred1

0      77.0
1      77.0
2      77.0
3      77.0
4      77.0
       ... 
434    79.0
435    78.0
436    77.0
437    78.0
438    75.0
Length: 439, dtype: float64

In [814]:
y_train

overall
509         79
510         79
511         79
513         79
514         79
...        ...
19234       47
19235       47
19236       47
19237       47
19238       47

[16668 rows x 1 columns]

##### Menggunakan algoritma Linear Regression

In [815]:
from sklearn.linear_model import LinearRegression
linreg = LinearRegression()
linreg = linreg.fit(X_train, y_train)

In [816]:
y_pred2 = linreg.predict(X_test)

In [817]:
y_pred2 = pd.DataFrame(y_pred2)
y_pred2

0
0    89.408260
1    86.452422
2    88.191532
3    86.361221
4    86.792015
..         ...
434  77.896423
435  77.590362
436  77.646146
437  76.432023
438  76.234326

[439 rows x 1 columns]

##### Menggunakan algoritma Random Forest Regressor

In [818]:
from sklearn.ensemble import RandomForestRegressor
rf = RandomForestRegressor()
rf = rf.fit(X_train, y_train)

C:\Users\CPU\AppData\Local\Temp\ipykernel_9404\3665841973.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  rf = rf.fit(X_train, y_train)


In [819]:
y_pred3 = rf.predict(X_test)

In [820]:
y_pred3 = pd.Series(y_pred3)

##### Menentukan algoritma yang paling baik dengan cara melakukan penilaian dengan parameter nilai error yang paling kecil diantara algoritma tersebut.

In [821]:
from sklearn.metrics import mean_absolute_error
mae1 = mean_absolute_error(y_test, y_pred1)
mae2 = mean_absolute_error(y_test, y_pred2)
mae3 = mean_absolute_error(y_test, y_pred3)

In [822]:
mae1

5.026195899772209

In [823]:
mae2

2.944567953587012

In [824]:
mae3

4.943745634016705

##### Setelah ditemukan algoritma dengan error terkecil (Linear Regression), y_pred2 disisipkan ke dalam dataframe test_dataset1 untuk selanjutnya diproses kembali.

In [825]:
new_dataset1 = test_dataset1.set_index('short_name').copy()

In [826]:
new_dataset1 = new_dataset1.reset_index()

In [827]:
update_dataset = pd.merge(new_dataset1,y_pred2,how="outer",left_index=True,right_index=True)

In [828]:
update_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 439 entries, 0 to 438
Data columns (total 7 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   short_name           439 non-null    object 
 1   overall              439 non-null    int64  
 2   passing              439 non-null    float64
 3   dribbling            439 non-null    float64
 4   movement_reactions   439 non-null    int64  
 5   mentality_composure  439 non-null    int64  
 6   0                    439 non-null    float64
dtypes: float64(3), int64(3), object(1)
memory usage: 24.1+ KB


In [829]:
update_dataset.rename(columns={0:"rating"},inplace=True)

In [830]:
import math

In [831]:
# Membulatkan nilai rating ke nilai puluhan terdekat
update_dataset['rating'] = update_dataset['rating'].apply(lambda x : math.ceil(x))

In [832]:
update_dataset

short_name  overall  passing  dribbling  movement_reactions  \
0             L. Messi       93     91.0       95.0                  94   
1       R. Lewandowski       92     79.0       86.0                  93   
2    Cristiano Ronaldo       91     80.0       88.0                  94   
3            Neymar Jr       91     86.0       94.0                  89   
4         K. De Bruyne       91     93.0       88.0                  91   
..                 ...      ...      ...        ...                 ...   
434          N. Vlašić       80     78.0       82.0                  78   
435        Pedro Porro       80     73.0       79.0                  78   
436       V. Tsygankov       80     78.0       80.0                  78   
437          V. Coufal       80     73.0       75.0                  78   
438            B. Saka       80     76.0       82.0                  76   

     mentality_composure  rating  
0                     96      90  
1                     88      87  
2                     95      89  
3                     93      87  
4                     89      87  
..                   ...     ...  
434                   77      78  
435                   78      78  
436                   76      78  
437                   72      77  
438                   73      77  

[439 rows x 7 columns]

##### Menggabungkan variabel wage_eur, value_eur, short_name, international_reputations, release_clause_eur, dari dataset pertama dengan variabel rating berdasarkan pemrosesan yang baru saja dilakukan menjadi satu dataframe baru.

In [833]:
first_data = data[['short_name', 'value_eur','international_reputation','release_clause_eur','wage_eur']].copy()

In [834]:
first_data

short_name    value_eur  international_reputation  \
0                L. Messi   78000000.0                         5   
1          R. Lewandowski  119500000.0                         5   
2       Cristiano Ronaldo   45000000.0                         5   
3               Neymar Jr  129000000.0                         5   
4            K. De Bruyne  125500000.0                         4   
...                   ...          ...                       ...   
19234           Song Defu      70000.0                         1   
19235           C. Porter     110000.0                         1   
19236            N. Logue     100000.0                         1   
19237           L. Rudden     110000.0                         1   
19238  E. Lalchhanchhuaha     110000.0                         1   

       release_clause_eur  wage_eur  
0             144300000.0  320000.0  
1             197200000.0  270000.0  
2              83300000.0  270000.0  
3             238700000.0  270000.0  
4             232200000.0  350000.0  
...                   ...       ...  
19234            114000.0    1000.0  
19235            193000.0     500.0  
19236            175000.0     500.0  
19237            239000.0     500.0  
19238            217000.0     500.0  

[19239 rows x 5 columns]

In [835]:
second_data = update_dataset[['short_name','rating']].copy()

In [836]:
second_data

short_name  rating
0             L. Messi      90
1       R. Lewandowski      87
2    Cristiano Ronaldo      89
3            Neymar Jr      87
4         K. De Bruyne      87
..                 ...     ...
434          N. Vlašić      78
435        Pedro Porro      78
436       V. Tsygankov      78
437          V. Coufal      77
438            B. Saka      77

[439 rows x 2 columns]

In [837]:
dataset2 = pd.merge(first_data,second_data,how='inner',left_on='short_name',right_on='short_name')

In [838]:
dataset2

short_name    value_eur  international_reputation  \
0             L. Messi   78000000.0                         5   
1       R. Lewandowski  119500000.0                         5   
2    Cristiano Ronaldo   45000000.0                         5   
3            Neymar Jr  129000000.0                         5   
4         K. De Bruyne  125500000.0                         4   
..                 ...          ...                       ...   
535          N. Vlašić   33500000.0                         2   
536        Pedro Porro   40500000.0                         1   
537       V. Tsygankov   33500000.0                         2   
538          V. Coufal   19000000.0                         1   
539            B. Saka   45500000.0                         2   

     release_clause_eur  wage_eur  rating  
0           144300000.0  320000.0      90  
1           197200000.0  270000.0      87  
2            83300000.0  270000.0      89  
3           238700000.0  270000.0      87  
4           232200000.0  350000.0      87  
..                  ...       ...     ...  
535          70400000.0   66000.0      78  
536                 NaN   80000.0      78  
537          80400000.0    1000.0      78  
538          36100000.0   73000.0      77  
539          95600000.0   50000.0      77  

[540 rows x 6 columns]

In [839]:
dataset2 = dataset2.dropna()

In [840]:
dataset2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 520 entries, 0 to 539
Data columns (total 6 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   short_name                520 non-null    object 
 1   value_eur                 520 non-null    float64
 2   international_reputation  520 non-null    int64  
 3   release_clause_eur        520 non-null    float64
 4   wage_eur                  520 non-null    float64
 5   rating                    520 non-null    int64  
dtypes: float64(3), int64(2), object(1)
memory usage: 28.4+ KB


In [841]:
scaling = dataset2.copy()

In [842]:
scaling['release_clause_eur'] = scaling['release_clause_eur'] / scaling['release_clause_eur'].max()
scaling['wage_eur'] = scaling['wage_eur'] /scaling['wage_eur'].max()
scaling['value_eur'] = scaling['value_eur'] /scaling['value_eur'].max()
scaling['international_reputation'] = scaling['international_reputation'] / scaling['international_reputation'].max()
scaling['rating'] = scaling['rating'] / scaling['rating'].max()

In [843]:
scaling

short_name  value_eur  international_reputation  \
0             L. Messi   0.402062                       1.0   
1       R. Lewandowski   0.615979                       1.0   
2    Cristiano Ronaldo   0.231959                       1.0   
3            Neymar Jr   0.664948                       1.0   
4         K. De Bruyne   0.646907                       0.8   
..                 ...        ...                       ...   
534            L. Díaz   0.001933                       0.2   
535          N. Vlašić   0.172680                       0.4   
537       V. Tsygankov   0.172680                       0.4   
538          V. Coufal   0.097938                       0.2   
539            B. Saka   0.234536                       0.4   

     release_clause_eur  wage_eur    rating  
0              0.386345  0.914286  1.000000  
1              0.527979  0.771429  0.966667  
2              0.223025  0.771429  0.988889  
3              0.639090  0.771429  0.966667  
4              0.621687  1.000000  0.966667  
..                  ...       ...       ...  
534            0.003213  0.001429  0.877778  
535            0.188487  0.188571  0.866667  
537            0.215261  0.002857  0.866667  
538            0.096653  0.208571  0.855556  
539            0.255957  0.142857  0.855556  

[520 rows x 6 columns]

##### Membagi data menjadi 2 bagian train dan test dengan ketentuan nilai rating lebih dari atau sama dengan 80 dijadikan test dataset.

In [844]:
train_dataset2 = scaling.loc[(scaling['rating'])<0.88888].copy()

In [845]:
train_dataset2

short_name  value_eur  international_reputation  release_clause_eur  \
42   G. Chiellini   0.061856                       0.8            0.053012   
50   K. Koulibaly   0.286082                       0.6            0.252744   
108  Luis Alberto   0.219072                       0.4            0.193574   
115      S. Savić   0.177835                       0.4            0.189290   
116      S. Savić   0.004639                       0.2            0.003213   
..            ...        ...                       ...                 ...   
534       L. Díaz   0.001933                       0.2            0.003213   
535     N. Vlašić   0.172680                       0.4            0.188487   
537  V. Tsygankov   0.172680                       0.4            0.215261   
538     V. Coufal   0.097938                       0.2            0.096653   
539       B. Saka   0.234536                       0.4            0.255957   

     wage_eur    rating  
42   0.251429  0.866667  
50   0.300000  0.877778  
108  0.271429  0.855556  
115  0.211429  0.877778  
116  0.008571  0.877778  
..        ...       ...  
534  0.001429  0.877778  
535  0.188571  0.866667  
537  0.002857  0.866667  
538  0.208571  0.855556  
539  0.142857  0.855556  

[222 rows x 6 columns]

In [846]:
X_train2 = train_dataset2[['value_eur','international_reputation','release_clause_eur','rating']]
y_train2 = train_dataset2[['wage_eur']]

In [847]:
test_dataset2 = scaling.loc[(scaling['rating'])>=0.888888].copy()

In [848]:
test_dataset2

short_name  value_eur  international_reputation  \
0             L. Messi   0.402062                       1.0   
1       R. Lewandowski   0.615979                       1.0   
2    Cristiano Ronaldo   0.231959                       1.0   
3            Neymar Jr   0.664948                       1.0   
4         K. De Bruyne   0.646907                       0.8   
..                 ...        ...                       ...   
493        V. Lindelöf   0.118557                       0.4   
504        I. Bennacer   0.154639                       0.4   
508         D. Zakaria   0.157216                       0.4   
511             Gerson   0.154639                       0.4   
515     Renato Sanches   0.170103                       0.6   

     release_clause_eur  wage_eur    rating  
0              0.386345  0.914286  1.000000  
1              0.527979  0.771429  0.966667  
2              0.223025  0.771429  0.988889  
3              0.639090  0.771429  0.966667  
4              0.621687  1.000000  0.966667  
..                  ...       ...       ...  
493            0.118608  0.300000  0.900000  
504            0.142704  0.117143  0.888889  
508            0.155288  0.080000  0.888889  
511            0.158768  0.117143  0.900000  
515            0.185542  0.108571  0.888889  

[298 rows x 6 columns]

In [849]:
X_test2 = test_dataset2[['value_eur','international_reputation','release_clause_eur','rating']]
y_test2 = test_dataset2[['wage_eur']]

##### Sama seperti proses sebelumnya, sekarang akan dilakukan pengujian dan penerapan algoritma supervised learning guna menemukan formula mana yang lebih baik.

##### Algoritma Decision Tree Regressor

In [850]:
from sklearn.tree import DecisionTreeRegressor

In [851]:
model2 = DecisionTreeRegressor()
model2 = model2.fit(X_train2, y_train2)

In [852]:
pred1 = model2.predict(X_test2)

##### Algoritma Linear Regression

In [853]:
from sklearn.linear_model import LinearRegression

In [854]:
linreg2 = LinearRegression()
linreg2 = linreg.fit(X_train2, y_train2)

In [855]:
pred2 = linreg2.predict(X_test2)

##### Algoritma Random Forest Regressor

In [856]:
from sklearn.ensemble import RandomForestRegressor

In [857]:
fr2 = RandomForestRegressor()
fr2 = fr2.fit(X_train2, y_train2)

C:\Users\CPU\AppData\Local\Temp\ipykernel_9404\2272597922.py:2: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  fr2 = fr2.fit(X_train2, y_train2)


In [858]:
pred3 = fr2.predict(X_test2)

##### Menghitung nilai error masing-masing algoritma dengan Mean Absolute Error

In [859]:
from sklearn.metrics import mean_absolute_error
mae_pred1 = mean_absolute_error(y_test2, pred1)
mae_pred2 = mean_absolute_error(y_test2, pred2)
mae_pred3 = mean_absolute_error(y_test2, pred3)

In [860]:
mae_pred1

0.11590316395014381

In [861]:
mae_pred2

0.09513710258680792

In [862]:
mae_pred3

0.11366855704697984

#### Karena algoritma LinReg memiliki nilai error paling kecil, maka data yang akan dipakai adalah prediksi dari Linreg

In [863]:
pred2 = pd.DataFrame(pred2)
pred2

0
0    0.519472
1    0.638198
2    0.400621
3    0.695137
4    0.632216
..        ...
293  0.158450
294  0.177666
295  0.183105
296  0.185281
297  0.248933

[298 rows x 1 columns]

In [864]:
test_dataset2 = test_dataset2.set_index('short_name').copy()

In [865]:
test_dataset2 = test_dataset2.reset_index()
test_dataset2

short_name  value_eur  international_reputation  \
0             L. Messi   0.402062                       1.0   
1       R. Lewandowski   0.615979                       1.0   
2    Cristiano Ronaldo   0.231959                       1.0   
3            Neymar Jr   0.664948                       1.0   
4         K. De Bruyne   0.646907                       0.8   
..                 ...        ...                       ...   
293        V. Lindelöf   0.118557                       0.4   
294        I. Bennacer   0.154639                       0.4   
295         D. Zakaria   0.157216                       0.4   
296             Gerson   0.154639                       0.4   
297     Renato Sanches   0.170103                       0.6   

     release_clause_eur  wage_eur    rating  
0              0.386345  0.914286  1.000000  
1              0.527979  0.771429  0.966667  
2              0.223025  0.771429  0.988889  
3              0.639090  0.771429  0.966667  
4              0.621687  1.000000  0.966667  
..                  ...       ...       ...  
293            0.118608  0.300000  0.900000  
294            0.142704  0.117143  0.888889  
295            0.155288  0.080000  0.888889  
296            0.158768  0.117143  0.900000  
297            0.185542  0.108571  0.888889  

[298 rows x 6 columns]

In [866]:
final_dataset = pd.merge(test_dataset2, pred2, how='inner', left_index=True, right_index=True)

In [867]:
final_dataset.rename(columns={0:"salary_in_euro"},inplace=True)

In [868]:
final_dataset

short_name  value_eur  international_reputation  \
0             L. Messi   0.402062                       1.0   
1       R. Lewandowski   0.615979                       1.0   
2    Cristiano Ronaldo   0.231959                       1.0   
3            Neymar Jr   0.664948                       1.0   
4         K. De Bruyne   0.646907                       0.8   
..                 ...        ...                       ...   
293        V. Lindelöf   0.118557                       0.4   
294        I. Bennacer   0.154639                       0.4   
295         D. Zakaria   0.157216                       0.4   
296             Gerson   0.154639                       0.4   
297     Renato Sanches   0.170103                       0.6   

     release_clause_eur  wage_eur    rating  salary_in_euro  
0              0.386345  0.914286  1.000000        0.519472  
1              0.527979  0.771429  0.966667        0.638198  
2              0.223025  0.771429  0.988889        0.400621  
3              0.639090  0.771429  0.966667        0.695137  
4              0.621687  1.000000  0.966667        0.632216  
..                  ...       ...       ...             ...  
293            0.118608  0.300000  0.900000        0.158450  
294            0.142704  0.117143  0.888889        0.177666  
295            0.155288  0.080000  0.888889        0.183105  
296            0.158768  0.117143  0.900000        0.185281  
297            0.185542  0.108571  0.888889        0.248933  

[298 rows x 7 columns]

In [869]:
final_dataset['release_clause_eur'] = final_dataset['release_clause_eur'] * dataset2['release_clause_eur'].max()
final_dataset['wage_eur'] = final_dataset['wage_eur'] * dataset2['wage_eur'].max()
final_dataset['value_eur'] = final_dataset['value_eur'] * dataset2['value_eur'].max()
final_dataset['international_reputation'] = final_dataset['international_reputation'] * dataset2['international_reputation'].max()
final_dataset['rating'] = final_dataset['rating'] * dataset2['rating'].max()
final_dataset['salary_in_euro'] = final_dataset['salary_in_euro'] * dataset2['wage_eur'].max()

In [870]:
final_dataset['release_clause_eur'] = final_dataset['release_clause_eur'].astype(int)
final_dataset['wage_eur'] = final_dataset['wage_eur'].astype(int)
final_dataset['value_eur'] = final_dataset['value_eur'].astype(int)
final_dataset['international_reputation'] = final_dataset['international_reputation'].astype(int)
final_dataset['rating'] = final_dataset['rating'].astype(int)
final_dataset['salary_in_euro'] = final_dataset['salary_in_euro'].astype(int)

In [871]:
final_dataset

short_name  value_eur  international_reputation  \
0             L. Messi   78000000                         5   
1       R. Lewandowski  119500000                         5   
2    Cristiano Ronaldo   45000000                         5   
3            Neymar Jr  129000000                         5   
4         K. De Bruyne  125500000                         4   
..                 ...        ...                       ...   
293        V. Lindelöf   23000000                         2   
294        I. Bennacer   30000000                         2   
295         D. Zakaria   30500000                         2   
296             Gerson   30000000                         2   
297     Renato Sanches   33000000                         3   

     release_clause_eur  wage_eur  rating  salary_in_euro  
0             144300000    320000      90          181815  
1             197200000    270000      87          223369  
2              83300000    270000      89          140217  
3             238700000    270000      87          243297  
4             232200000    350000      87          221275  
..                  ...       ...     ...             ...  
293            44300000    105000      81           55457  
294            53300000     41000      80           62183  
295            57999999     28000      80           64086  
296            59300000     41000      81           64848  
297            69300000     38000      80           87126  

[298 rows x 7 columns]

In [872]:
overprice = final_dataset.loc[(final_dataset['salary_in_euro'])<(final_dataset['wage_eur'])].copy()

In [873]:
overprice

short_name  value_eur  international_reputation  \
0             L. Messi   78000000                         5   
1       R. Lewandowski  119500000                         5   
2    Cristiano Ronaldo   45000000                         5   
3            Neymar Jr  129000000                         5   
4         K. De Bruyne  125500000                         4   
..                 ...        ...                       ...   
289          Q. Promes   20500000                         3   
290            D. Alli   26499999                         3   
291         C. Tolisso   24500000                         3   
292           N. Keïta   27500000                         2   
293        V. Lindelöf   23000000                         2   

     release_clause_eur  wage_eur  rating  salary_in_euro  
0             144300000    320000      90          181815  
1             197200000    270000      87          223369  
2              83300000    270000      89          140217  
3             238700000    270000      87          243297  
4             232200000    350000      87          221275  
..                  ...       ...     ...             ...  
289            32800000     68000      80           67056  
290            52300000     89999      81           77989  
291            42300000     78000      80           72737  
292            52900000    110000      80           60513  
293            44300000    105000      81           55457  

[146 rows x 7 columns]

In [874]:
underprice = final_dataset.loc[(final_dataset['salary_in_euro'])>(final_dataset['wage_eur'])].copy()

In [875]:
underprice

short_name  value_eur  international_reputation  release_clause_eur  \
5         K. Mbappé  194000000                         4           373500000   
14       J. Kimmich  108000000                         4           186300000   
16        L. Suárez   44500000                         5            91200000   
17        L. Suárez   23500000                         1            54100000   
18        L. Suárez     850000                         1             1300000   
..              ...        ...                       ...                 ...   
288           Roger     550000                         1              784000   
294     I. Bennacer   30000000                         2            53300000   
295      D. Zakaria   30500000                         2            57999999   
296          Gerson   30000000                         2            59300000   
297  Renato Sanches   33000000                         3            69300000   

     wage_eur  rating  salary_in_euro  
5      230000      87          311309  
14     160000      85          193690  
16     135000      86          140715  
17      23000      86           44560  
18       3000      86           12724  
..        ...     ...             ...  
288      2000      80            8605  
294     41000      80           62183  
295     28000      80           64086  
296     41000      81           64848  
297     38000      80           87126  

[152 rows x 7 columns]

#### KESIMPULAN
##### Ada 146 pemain yang mendapatkan gaji lebih tinggi dari yang seharusnya diterima atas pertimbangan rating, reputasi internasional, harga jual, dan nilai jual dalam kontrak. Sementara itu ada 152 pemain dengan besaran gaji lebih rendah dibandingkan kontribusinya atas pertimbangan faktor-faktor tadi.